packages/components/src/box-control/index.js

```diff
@@ -120,12 +120,9 @@ export default function BoxControl( {
 		<Root id={ id } role="group" aria-labelledby={ headingId }>
 			<Header className="component-box-control__header">
 				<FlexItem>
-					<Text
-						id={ headingId }
-						className="component-box-control__label"
```

Comment:
I take it we are ok removing the `components-box-control__label` class here given the `BoxControl` is still exported as experimental?

I ask, as later in this PR, we keep the `components-form-token-field__label` class desite removing the styles for it.

Reply:
Great catch! I am actually quite inclined to remove unused CSS classes _even if_ they are in stable components, unless it's a class that we can imagine a lot of third-party consumers to be overriding in a crucial way (e.g. `components-popover__content`). (@ciampo Does that sound about right or should we be more conservative about it?)

The reason I left `components-form-token-field__label` intact is more because it's being used as a selector in an old e2e test 🙃

Reply:
Sorry for the late answer, I'm currently a bit behind on my review queue 😅 

I believe we should address this situation in a more general way (and potentially even add a paragraph about the strategy we decide to adopt to the contributors' guidelines).

In general. I'd _love_ to remove hardcoded classnames — in practice, it may cause breakage for consumers that rely on them.

Should we consider the removal of hardcoded classnames a breaking change?

- On one hand, hardcoded classnames are not part of the public APIs of the components (except for a [these 4 components](https://cs.github.com/?scopeName=All+repos&scope=&q=%22components-%22+repo%3AWordPress%2Fgutenberg+path%3A%2Fpackages%2Fcomponents%2Fsrc%2F**%2F*.md)) — and so, technically it would not be a breaking change
- On the other hand, hardcoded classnames have been used outside of the package for years (both inside and outside the Gutenberg repo).

For example, even in the case of the removal `component-box-control__label`, we're not 100% guaranteed that it won't cause breakage to some external consumers.

In general, I believe we should _at least_ aim at removing all instances of `component-*` classnames outside of the components' package. After that, depending on how conservative we want to be, we could:

- also delete the hardcoded classnames in the source components
- keep them as "legacy" support, but add a ESLINT rule forbidding their usage in Gutenberg (although this won't stop external consumers from using them)

What do folks think?

Reply:
> I believe we should address this situation in a more general way (and potentially even add a paragraph about the strategy we decide to adopt to the contributors' guidelines).

Absolutely, I don't think it's explicitly written anywhere.

> Should we consider the removal of hardcoded classnames a breaking change?

Could it cause breakage? Yes. Should _we_ consider it a Breaking Change? No.

I was a consumer of the wp-components package before coming to the maintainer side, and I was never under the impression that CSS classnames (or styling, or DOM structure, for that matter) were guaranteed not to change. I think most people realize that overriding internal styling is fragile, and that it's their responsibility to protect against breakage if they are going to do so. Unless these classnames are publicly documented, as in the examples you raised 😅

Reply:
Personally, I agree with the notion that while it might cause breakage, we shouldn't necessarily consider it a breaking change. That said, I think communication and documentation will be key to maintaining as much good will as possible.

Reply:
Agree with both of you. I opened https://github.com/WordPress/gutenberg/issues/43083 to keep track of this task separately

Reply:
For reference, the project does have existing documentation about backwards compatibility around [class names and DOM updates](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/backward-compatibility.md#class-names-and-dom-updates).

> Class names and DOM nodes used inside the tree of React components are not considered part of the public API and can be modified.
> 
> Changes to these should be done with caution as it can affect the styling and behavior of third-party code (Even if they should not rely on these in the first place). Keep the old ones if possible. If not, document the changes and write a dev note.

Reply:
Thank you @nerrad ! Seems to align with our thoughts above